load_data.py

```diff
@@ -1,28 +1,132 @@
-def load_listenbrainz_dump():
-    """ Returns an RDD of format (user_name, track_name, release_name, timestamp, additional_info)
-    """
-    raise NotImplementedError
+import json
+import os
+import sys
+from collections import defaultdict
+from pyspark import SparkConf, SparkContext
+from pyspark.sql import SparkSession, Row
 
 
-def prepare_user_data():
-    """ Returns an RDD that is of the form (user_id, user_name)
-    """
-    raise NotImplementedError
+IMPORT_CHUNK_SIZE = 100000
+WRITE_RDDS_TO_DISK = True
+
+spark = SparkSession \
+        .builder \
+        .appName("LB recommender") \
+        .getOrCreate()
 
 
-def prepare_recording_data():
-    """ Returns an RDD of the form (recording_id, recording_name)
+def load_listenbrainz_dump(directory, sc):
     """
-    raise NotImplementedError
+    Reads files from an uncompressed ListenBrainz dump and creates user and listens RDDs.
 
+    Args:
+        directory: Path of the directory containing the LB dump files.
+        sc: The SparkContext object.
 
-def prepare_listen_counts():
-    """ calculates ratings, returns rdd of the form (user_id, recording_id, play_count)
+    Returns:
+        users_df: DataFrame containing user info (user_id, user_name).
+        listens_df: DataFrame containing ALL listens with user-names.
     """
-    raise NotImplementedError
+    listens_rdd = sc.parallelize([])
+    users = list()
+    with open(os.path.join(directory, 'index.json')) as f:
+        index = json.load(f)
+    file_contents = defaultdict(list)
+    for i, (user, info) in enumerate(index.items()):
+        file_contents[info['file_name']].append({
+            'user_name': user,
+            'offset': info['offset'],
+            'size': info['size'],
+        })
+        users.append((i+1, user))
+
+    users_rdd = sc.parallelize(users)
+
+    for file_name in file_contents:
+        file_contents[file_name] = sorted(file_contents[file_name], key=lambda x: x['offset'])
+        file_path = os.path.join(directory, 'listens', file_name[0], file_name[0:2], '%s.listens' % file_name)
+        with open(file_path, 'r') as f:
+            for user in file_contents[file_name]:
+                print('Importing user %s...' % user['user_name'])
+                assert(f.tell() == user['offset'])
+                bytes_read = 0
+                listens = []
+                while bytes_read < user['size']:
+                    line = f.readline()
+                    bytes_read += len(line)
+                    listen = json.loads(line)
+                    listen["user_name"] = user['user_name']
+                    listens.append(listen)
+                    if len(listens) > IMPORT_CHUNK_SIZE:
+                        listens_rdd = listens_rdd.union(sc.parallelize(listens))
+                        listens = []
+                    if len(listens) > 0:
```

Comment:
Shouldn't these two lines be indented one level less, so they only add the listens that are left after a users listens have been read completely, but the last batch didn't include `IMPORT_CHUNK_SIZE` listens? At the moment, it looks like after reading one line, all the lines that have been read for the current batch will be added to `listens_rdd`, over and over again (`RDD.union` does not eliminate duplicates).

Reply:
Looks to be a problem yes. Fortunately this code is already out of date. :)